.github/workflows/test.yml

```diff
@@ -1,81 +1,293 @@
 name: Test
 
 on:
-  push:
   pull_request:
+    branches:
+      - master
+  push:
+    branches:
+      - master
+
+env:
+  REGISTRY: ghcr.io
 
 jobs:
+  # Apply formatter and fail on errors
   check_format:
-    if: '! github.event.pull_request.draft'
     name: Check code formatting
     runs-on: ubuntu-22.04
     steps:
       - name: Checkout repo
         uses: actions/checkout@v3
-        with:
-          submodules: recursive
 
       - name: Setup Python
         uses: actions/setup-python@v4
         with:
-          python-version: '3.10'
+          python-version: "3.10"
+
       - name: Check code formatting
-        uses: pre-commit/action@v2.0.3
+        uses: pre-commit/action@v3.0.0
+
+  # Build all images for services declaring a build
+  # Push them onto the repository's Github Container Registry
+  build_push__app:
```

Comment:
Can we put that all into a matrix?

```
jobs:
  build:
    runs-on: ubuntu-22.04
    strategy:
      matrix:
        include:
          - image_name: 'app'
            build_context: './docker-app'
            target: 'webserver_test'
          - image_name: ''worker_wrapper
            ...
```

Reply:
What steps would be associated with `build`?  Something like:

```
uses: docker/build-push-action@v4
        with:
          context: ./docker-app
          target: webserver_test
          build-args: DEBUG_BUILD=1
          tags: ${{ env.REGISTRY }}/${{ github.repository }}/app:${{ github.head_ref }}.${{ github.sha }}
          cache-from: type=gha
          cache-to: type=gha,mode=max
          push: true
```
? If yes, every iteration has to fetch several values  -- a different numbers of values depending on the image -- so that  a combination of `context`, `build-args` and `tags` are well-defined. Isn't it more complicated than what is done now?

Reply:
Fair question, I think we can deal with context and tags, but build-args and target is complicated...